tensorflow/core/kernels/depthtospace_op.cc

```diff
@@ -112,23 +113,6 @@ class DepthToSpaceOp : public OpKernel {
     auto Tinput = input.tensor<T, kDims>();
     auto Toutput = outputs_tensor->tensor<T, kDims>();
 
-    if (std::is_same<Device, GPUDevice>::value) {
-      if (is_int8x4) {
-        // NCHW_VECT_C with 4 x qint8 can be treated as NCHW int32.
-        auto Tinput_v = input.template reinterpret_last_dimension<int32, 4>();
-        auto Toutput_v = outputs_tensor->reinterpret_last_dimension<int32, 4>();
-        functor::DepthToSpaceOpFunctor<GPUDevice, int32, FORMAT_NCHW> functor;
-        functor(context->eigen_device<GPUDevice>(), Tinput_v, block_size_,
-                Toutput_v);
-        return;
-      } else if (data_format_ == FORMAT_NCHW) {
-        functor::DepthToSpaceOpFunctor<GPUDevice, T, FORMAT_NCHW> functor;
-        functor(context->eigen_device<GPUDevice>(), Tinput, block_size_,
-                Toutput);
-        return;
-      }
-    }
-
     // NOTE: Assumes data_format_ == FORMAT_NHWC here, since we have rejected
     // (CPU && data_format_ != FORMAT_NHWC) in the constructor.
 
```

Comment:
Can we specialize just this part that's different between GPU and CPU?  Same comment for `SpaceToDepthOp`.

Reply:
That's what I tried to do initially - specialize only the Compute() function, but [AFAIU ](https://stackoverflow.com/a/15374950/1675506) there is no way to specialize part of the class definition only: The template signature of the function has to match that of the class, so to partially specialize the function you need to partially specialize the class. 

Taking the GPU-specialized code into a separate function would result in the same situation - the class still needs to be specialized.

One could probably get around this with some inheritance trickery, but this IMO is making this messier than it should be. Not that I like this code duplication much more...

Reply:
I mean something like:

```
template <class Device>
struct DepthToSpaceFunctorWrapper {
  void operator(params) {
    // .. generic implementation
  }
};

template <>
struct DepthToSpaceFunctorWrapper<GPUDevice> {
  void operator(params) {
    // .. GPU implementation
  }
};

void Compute(...) {
  // Common stuff ..
  auto Toutput = outputs_tensor->tensor<T, kDims>();
  DepthToSpaceFunctorWrapper<Device>{}(params as needed);
}
```

Reply:
Wait, isn't there a much simpler solution? In the if clause [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/depthtospace_op.cc#L115), if one is inside it, it is guaranteed that Device _is_ GPUDevice. So why not replace `functor::DepthToSpaceOpFunctor<GPUDevice, T, FORMAT_NCHW> functor;` by `functor::DepthToSpaceOpFunctor<`**Device**`, T, FORMAT_NCHW> functor;` inside this clause? This should be identical, no?